localstack-core/localstack/aws/handlers/validation.py

```diff
@@ -0,0 +1,86 @@
+
+LOG = logging.getLogger(__name__)
+
+
+# TODO: replace with from importlib.resources.files when https://github.com/python/importlib_resources/issues/311 is
+#   resolved. Import from a namespace package is broken when installing in editable mode.
+oas_path = os.path.join(os.path.dirname(__file__), "..", "..", "openapi.yaml")
+
+
+class OpenAPIValidator(Handler):
+    openapi: "OpenAPI"
+
+    def __init__(self) -> None:
+        path = Path(oas_path)
+        assert path.exists()
+        self.openapi = OpenAPI.from_path(path)
+
+
+class OpenAPIRequestValidator(OpenAPIValidator):
+    """
+    Validates the requests to the LocalStack public endpoints (the ones with a _localstack or _aws prefix) against
+    a OpenAPI specification.
+    """
+
+    def __call__(self, chain: HandlerChain, context: RequestContext, response: Response):
+        if not config.OPENAPI_VALIDATE_REQUEST:
+            return
+
+        path = context.request.path
+
+        if path.startswith(f"{INTERNAL_RESOURCE_PATH}/") or path.startswith("/_aws/"):
+            try:
+                self.openapi.validate_request(WerkzeugOpenAPIRequest(context.request))
+            except RequestValidationError as e:
+                # Note: in this handler we only check validation errors, e.g., wrong body, missing required in the body.
+                response.status_code = 400
+                response.set_json({"error": "Bad Request", "message": str(e)})
+                chain.stop()
+            except OpenAPIError:
+                # Other errors can be raised when validating a request against the OpenAPI specification.
+                #   The most common are: ServerNotFound, OperationNotFound, or PathNotFound.
+                #   We explicitly do not check any other error but RequestValidationError ones.
+                #   We shallow the exception to avoid excessive logging (e.g., a lot of ServerNotFound), as the only
+                #   purpose of this handler is to check for request validation errors.
+                pass
+
+
+class OpenAPIResponseValidator(OpenAPIValidator):
+    def __call__(self, chain: HandlerChain, context: RequestContext, response: Response):
+        # We are more lenient in validating the responses. The use of this flag is intended for test.
+        if not config.OPENAPI_VALIDATE_RESPONSE:
+            return
+
+        path = context.request.path
+
+        if path.startswith(f"{INTERNAL_RESOURCE_PATH}/") or path.startswith("/_aws/"):
+            try:
+                self.openapi.validate_response(
+                    WerkzeugOpenAPIRequest(context.request),
+                    WerkzeugOpenAPIResponse(response),
+                )
+            except OpenAPIError as exc:
+                LOG.error("Response validation failed for %s: $s", path, exc)
+                response.status_code = 500
+                response.set_json({"error": exc.__class__.__name__, "message": str(exc)})
+                chain.stop()
```

Comment:
q: Here we are (other than in the request validator) way stricter. This can make sense (because this is deactivated by default), but:
- Is this on purpose? Do we really raise on every `OpenAPIError` here?
- If it's on purpose, it would make sense to explain the reasoning here in a comment :)

Reply:
The idea was to be stricter, as you suggested, since errors here should be implementation errors on the LS side.
However, I don't yet have a precise idea of every possible error we might encounter here. Therefore, I followed the same logic of the request validation and only checked for `ResponseValidationError` at the moment. Thoughts?